R/tree.R

```diff
@@ -0,0 +1,314 @@
+
+  depth <- length(branch_hist)
+
+  # Build branch string from branch history
+  # Start with empty spaces
+  branch_chars <- rep_len("  ", depth)
+
+  # Store history in short name so logic is more legible
+  branch_chars[branch_hist == "child"] <- paste0(opts$vertical, " ")
+  branch_chars[branch_hist == "pre-attrs"] <- paste0(opts$vertical_attr, " ")
+
+  # Next update the final element (aka the current step) with the correct branch type
+  last_step <- branch_hist[depth]
+  root_node <- length(branch_hist) == 0
+  branch_chars[depth] <- if(root_node) "" else paste0(
+    if(grepl("last", last_step)) opts$final_branch else opts$branch,
+    if(grepl("attribute", last_step)) opts$horizontal_attr else opts$horizontal
+  )
+
+  # Build label
+  label <- paste0(
+    x_id,
+    if(!is.null(x_id) && x_id != "") ":",
+    tree_label(x,
+               class_printer = opts$class_printer,
+               val_printer = opts$val_printer,
+               remove_newlines = opts$remove_newlines)
+  )
+
+  # Do the actual printing to the console
+  cat(paste(branch_chars, collapse = ""), label, "\n", sep = "")
+
+  x_attributes <- attributes(x)
+  if(attr_mode){
+    # Filter out "names" attribute as this is already shown by tree
+    x_attributes <- x_attributes[names(x_attributes) != "names"]
+  }
+  has_attributes <- length(x_attributes) > 0 & opts$show_attributes
+
+  # ===== Start recursion logic
+  # Turn into a s3 method for recursion
+  if(!is.atomic(x) & depth <= opts$max_depth & !is.function(x) & !is.environment(x)){
+    children <- as.list(x)
+
+    # Traverse children, if any exist
+    n_children <- length(children)
+
+    # If children have names, give them the names
+    for (i in seq_along(children)) {
+      id <- names(x)[i]
+      if(is.null(id) & opts$index_arraylike) id <- i
+
+      child_type <- if(i < n_children){
+        "child"
+      } else if(has_attributes) {
+        "pre-attrs"
+      } else {
+        "last-child"
+      }
+      tree_internal(
+        x = children[[i]],
+        x_id = id,
+        branch_hist = c(branch_hist, child_type),
+        opts = opts
+      )
+    }
+  }
+  # ===== End recursion logic
+
+  # Add any attributes as an "attr" prefixed children at end
+  if(has_attributes){
+    n_attributes <- length(x_attributes)
+    for(i in seq_len(n_attributes)){
+      tree_internal(
+        x = x_attributes[[i]],
+        x_id = crayon::italic(paste0("<attr>", names(x_attributes)[i])),
+        opts = opts,
+        branch_hist = c(branch_hist, paste0(if(i == n_attributes) "last-", "attribute")),
+        attr_mode = TRUE # Let tree know this is an attribute
+      )
+    }
+  }
+}
+
+#' Build element or node label in tree
+#'
+#' These methods control how the value of a given node is printed. New methods
+#' can be added if support is needed for a novel class
+#'
+#' @inheritParams tree
+#'
+#' @export
+tree_label <- function(x, val_printer, class_printer, remove_newlines){
+  UseMethod("tree_label")
```

Comment:
Hmmmm, it's not clear to me that you actually want to use S3 dispatch at all here.

Reply:
The motivation for using S3 dispatch was to enable external developers to hook up custom printing functions for their classes. For instance maybe `Shiny.tag` wants to print as `<h1/>` etc..

Without that motivation, it does feel like creating unnecessary namespace exports. 

It also feels like if the goal is truly to most accurately represent the structure then we wouldn't want custom printing for classes. _However_, if we wanted to have the function serve as a base for different tree-printing applications then maybe we should either keep the S3 dispatch or come up with another customization interface.